presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.ApplyNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.Map;
+
+public class VerifyNoApplyNodeLeftPlanOptimizer
```

Comment:
Sounds ok. Thanks for explanation.

śr., 4.05.2016, 21:46 użytkownik Martin Traverso <notifications@github.com>
napisał:

> In
> presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java
> <https://github.com/prestodb/presto/pull/4403#discussion_r62102119>:
>
> > +package com.facebook.presto.sql.planner.optimizations;
> > +
> > +import com.facebook.presto.Session;
> > +import com.facebook.presto.spi.PrestoException;
> > +import com.facebook.presto.spi.StandardErrorCode;
> > +import com.facebook.presto.spi.type.Type;
> > +import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
> > +import com.facebook.presto.sql.planner.SimplePlanVisitor;
> > +import com.facebook.presto.sql.planner.Symbol;
> > +import com.facebook.presto.sql.planner.SymbolAllocator;
> > +import com.facebook.presto.sql.planner.plan.ApplyNode;
> > +import com.facebook.presto.sql.planner.plan.PlanNode;
> > +
> > +import java.util.Map;
> > +
> > +public class VerifyNoApplyNodeLeftPlanOptimizer
>
> Generally, the way this ends up being used is:
>
>    - Get a "broken plan error"
>    - Distill the query to a small reproducible case
>    - Fire up Presto in a debugger & trace. This commonly involves adding
>    a call to validate() before/after each optimizer.
>
> So, you almost always have to tweak the code to use track down issues, so
> I think it's acceptable to just validate at the end and leave it up to the
> developer to insert calls and enable/disable checkers as necessary while
> debugging.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/prestodb/presto/pull/4403/files/745f4349fafc47aa5051fb3a577af7332ddd11ef#r62102119>
>